File: IAIso-v5.0/core/iaiso-ruby/test/consent_test.rb
# frozen_string_literal: true

require_relative "test_helper"
require "iaiso/consent"

class ConsentTest < Minitest::Test
  SECRET = "test_secret_long_enough_for_hs256_security_xx"

  def test_scope_exact_match
    assert IAIso::Consent::Scopes.granted(["tools.search"], "tools.search")
  end

  def test_scope_prefix_at_boundary
    assert IAIso::Consent::Scopes.granted(["tools"], "tools.search")
  end

  def test_scope_substring_not_boundary
    refute IAIso::Consent::Scopes.granted(["tools"], "toolsbar")
  end

  def test_scope_more_specific_doesnt_satisfy
    refute IAIso::Consent::Scopes.granted(["tools.search.bulk"], "tools.search")
  end

  def test_empty_requested_raises
    assert_raises(ArgumentError) { IAIso::Consent::Scopes.granted(["tools"], "") }
  end

  def test_issue_verify_roundtrip
    issuer = IAIso::Consent::Issuer.new(
      algorithm: IAIso::Consent::Algorithm::HS256, issuer: "iaiso",
      hs_key: SECRET, clock: ->{ 1_700_000_000 },
    )
    scope = issuer.issue(subject: "user-1", scopes: ["tools.search", "tools.fetch"])
    refute_empty scope.token

    verifier = IAIso::Consent::Verifier.new(
      algorithm: IAIso::Consent::Algorithm::HS256, issuer: "iaiso",
      hs_key: SECRET, clock: ->{ 1_700_000_001 },
    )
    verified = verifier.verify(scope.token)
    assert_equal "user-1", verified.subject
    assert verified.grants?("tools.search")
  end

  def test_verify_rejects_expired
    issuer = IAIso::Consent::Issuer.new(hs_key: SECRET, clock: ->{ 1_700_000_000 })
    scope = issuer.issue(subject: "u", scopes: ["tools"], ttl_seconds: 1)
    verifier = IAIso::Consent::Verifier.new(hs_key: SECRET, clock: ->{ 1_700_000_010 })
    assert_raises(IAIso::Consent::ExpiredTokenError) { verifier.verify(scope.token) }
  end

  def test_verify_honors_revocation
    issuer = IAIso::Consent::Issuer.new(hs_key: SECRET, clock: ->{ 1_700_000_000 })
    scope = issuer.issue(subject: "u", scopes: ["tools"])
    rl = IAIso::Consent::RevocationList.new
    rl.revoke(scope.jti)
    verifier = IAIso::Consent::Verifier.new(
      hs_key: SECRET, revocation_list: rl, clock: ->{ 1_700_000_001 },
    )
    assert_raises(IAIso::Consent::RevokedTokenError) { verifier.verify(scope.token) }
  end

  def test_verify_honors_execution_binding
    issuer = IAIso::Consent::Issuer.new(hs_key: SECRET, clock: ->{ 1_700_000_000 })
    scope = issuer.issue(subject: "u", scopes: ["tools"], execution_id: "exec-abc")
    verifier = IAIso::Consent::Verifier.new(hs_key: SECRET, clock: ->{ 1_700_000_001 })
    assert_raises(IAIso::Consent::InvalidTokenError) do
      verifier.verify(scope.token, requested_execution_id: "exec-xyz")
    end
  end

  def test_verify_rejects_tampered_token
    issuer = IAIso::Consent::Issuer.new(hs_key: SECRET, clock: ->{ 1_700_000_000 })
    scope = issuer.issue(subject: "u", scopes: ["tools"])
    tampered = scope.token[0...-5] + "XXXXX"
    verifier = IAIso::Consent::Verifier.new(hs_key: SECRET, clock: ->{ 1_700_000_001 })
    assert_raises(IAIso::Consent::InvalidTokenError) { verifier.verify(tampered) }
  end

  def test_generate_hs256_secret_length
    s = IAIso::Consent::Issuer.generate_hs256_secret
    assert s.length >= 64
  end
end

File: IAIso-v5.0/core/iaiso-php/tests/Unit/Consent/ConsentTest.php
<?php

declare(strict_types=1);

namespace IAIso\Tests\Unit\Consent;

use IAIso\Consent\Algorithm;
use IAIso\Consent\ExpiredTokenException;
use IAIso\Consent\InvalidTokenException;
use IAIso\Consent\Issuer;
use IAIso\Consent\RevocationList;
use IAIso\Consent\RevokedTokenException;
use IAIso\Consent\Scopes;
use IAIso\Consent\Verifier;
use PHPUnit\Framework\TestCase;

final class ConsentTest extends TestCase
{
    private const SECRET = 'test_secret_long_enough_for_hs256_security_xx';

    public function testScopeExactMatch(): void
    {
        self::assertTrue(Scopes::granted(['tools.search'], 'tools.search'));
    }

    public function testScopePrefixAtBoundary(): void
    {
        self::assertTrue(Scopes::granted(['tools'], 'tools.search'));
    }

    public function testScopeSubstringNotBoundary(): void
    {
        self::assertFalse(Scopes::granted(['tools'], 'toolsbar'));
    }

    public function testScopeMoreSpecificDoesntSatisfyLessSpecific(): void
    {
        self::assertFalse(Scopes::granted(['tools.search.bulk'], 'tools.search'));
    }

    public function testEmptyRequestedThrows(): void
    {
        $this->expectException(\InvalidArgumentException::class);
        Scopes::granted(['tools'], '');
    }

    public function testIssueVerifyRoundtrip(): void
    {
        $issuer = Issuer::builder()
            ->hsKey(self::SECRET)->algorithm(Algorithm::HS256)
            ->issuer('iaiso')->clock(fn() => 1_700_000_000)->build();
        $scope = $issuer->issue('user-1', ['tools.search', 'tools.fetch']);
        self::assertNotEmpty($scope->token);

        $verifier = Verifier::builder()
            ->hsKey(self::SECRET)->algorithm(Algorithm::HS256)
            ->issuer('iaiso')->clock(fn() => 1_700_000_001)->build();
        $verified = $verifier->verify($scope->token);
        self::assertSame('user-1', $verified->subject);
        self::assertTrue($verified->grants('tools.search'));
    }

    public function testVerifyRejectsExpired(): void
    {
        $issuer = Issuer::builder()->hsKey(self::SECRET)
            ->clock(fn() => 1_700_000_000)->build();
        $scope = $issuer->issue('u', ['tools'], null, 1);

        $verifier = Verifier::builder()->hsKey(self::SECRET)
            ->clock(fn() => 1_700_000_010)->build();
        $this->expectException(ExpiredTokenException::class);
        $verifier->verify($scope->token);
    }

    public function testVerifyHonorsRevocation(): void
    {
        $issuer = Issuer::builder()->hsKey(self::SECRET)
            ->clock(fn() => 1_700_000_000)->build();
        $scope = $issuer->issue('u', ['tools']);
        $rl = new RevocationList();
        $rl->revoke($scope->jti);

        $verifier = Verifier::builder()->hsKey(self::SECRET)
            ->revocationList($rl)->clock(fn() => 1_700_000_001)->build();
        $this->expectException(RevokedTokenException::class);
        $verifier->verify($scope->token);
    }

    public function testVerifyHonorsExecutionBinding(): void
    {
        $issuer = Issuer::builder()->hsKey(self::SECRET)
            ->clock(fn() => 1_700_000_000)->build();
        $scope = $issuer->issue('u', ['tools'], 'exec-abc');

        $verifier = Verifier::builder()->hsKey(self::SECRET)
            ->clock(fn() => 1_700_000_001)->build();
        $this->expectException(InvalidTokenException::class);
        $verifier->verify($scope->token, 'exec-xyz');
    }

    public function testVerifyRejectsTamperedToken(): void
    {
        $issuer = Issuer::builder()->hsKey(self::SECRET)
            ->clock(fn() => 1_700_000_000)->build();
        $scope = $issuer->issue('u', ['tools']);
        $tampered = substr($scope->token, 0, -5) . 'XXXXX';

        $verifier = Verifier::builder()->hsKey(self::SECRET)
            ->clock(fn() => 1_700_000_001)->build();
        $this->expectException(InvalidTokenException::class);
        $verifier->verify($tampered);
    }

    public function testGenerateHs256SecretLength(): void
    {
        $s = Issuer::generateHs256Secret();
        self::assertGreaterThanOrEqual(64, strlen($s));
    }
}
